modules/consensus/database.go

```diff
@@ -56,417 +86,254 @@ func openDB(filename string) (*setDB, error) {
+	return nil
 }
 
-// inBucket checks if an item with the given key is in the bucket
-func (db *setDB) inBucket(bucket []byte, key interface{}) bool {
-	exists, err := db.Exists(bucket, encoding.Marshal(key))
+// blockHeight returns the height of the blockchain.
+func blockHeight(tx *bolt.Tx) types.BlockHeight {
+	var height int
+	bh := tx.Bucket(BlockHeight)
+	err := encoding.Unmarshal(bh.Get(BlockHeight), &height)
 	if build.DEBUG && err != nil {
 		panic(err)
 	}
-	return exists
+	if height < 0 {
+		panic(height)
+	}
+	return types.BlockHeight(height)
 }
 
-// lenBucket is a simple wrapper for bucketSize that panics on error
-func (db *setDB) lenBucket(bucket []byte) uint64 {
-	s, err := db.BucketSize(bucket)
+// currentBlockID returns the id of the most recent block in the consensus set.
+func currentBlockID(tx *bolt.Tx) types.BlockID {
+	id, err := getPath(tx, blockHeight(tx))
 	if build.DEBUG && err != nil {
 		panic(err)
 	}
-	return s
+	return id
 }
 
-// forEachItem runs a given function on every element in a given
-// bucket name, and will panic on any error
-func (db *setDB) forEachItem(bucket []byte, fn func(k, v []byte) error) {
-	err := db.View(func(tx *bolt.Tx) error {
-		b := tx.Bucket(bucket)
-		if build.DEBUG && b == nil {
-			panic(errNilBucket)
-		}
-		return b.ForEach(fn)
-	})
-	if err != nil {
+// currentProcessedBlock returns the most recent block in the consensus set.
+func currentProcessedBlock(tx *bolt.Tx) *processedBlock {
+	pb, err := getBlockMap(tx, currentBlockID(tx))
+	if build.DEBUG && err != nil {
 		panic(err)
 	}
+	return pb
 }
 
-// pushPath inserts a block into the database at the "end" of the chain, i.e.
-// the current height + 1.
-//
-// DEPRECATED
-func (db *setDB) pushPath(bid types.BlockID) error {
-	value := encoding.Marshal(bid)
-	return db.Update(func(tx *bolt.Tx) error {
-		b := tx.Bucket(BlockPath)
-		key := encoding.EncUint64(uint64(b.Stats().KeyN))
-		return b.Put(key, value)
-	})
-}
-
-// popPath removes a block from the "end" of the chain, i.e. the block
-// with the largest height.
-func (db *setDB) popPath() error {
-	return db.Update(func(tx *bolt.Tx) error {
-		b := tx.Bucket(BlockPath)
-		key := encoding.EncUint64(uint64(b.Stats().KeyN - 1))
-		return b.Delete(key)
-	})
-}
+// getBlockMap returns a processed block with the input id.
+func getBlockMap(tx *bolt.Tx, id types.BlockID) (*processedBlock, error) {
+	// Look up the encoded block.
+	pbBytes := tx.Bucket(BlockMap).Get(id[:])
+	if pbBytes == nil {
+		return nil, errNilItem
+	}
 
-// getPath retreives the block id of a block at a given hegiht from the path
-//
-// DEPRECATED
-func (db *setDB) getPath(h types.BlockHeight) (id types.BlockID) {
-	idBytes, err := db.getItem(BlockPath, h)
-	if err != nil {
+	// Decode the block - should never fail.
+	var pb processedBlock
+	err := encoding.Unmarshal(pbBytes, &pb)
+	if build.DEBUG && err != nil {
 		panic(err)
 	}
-	err = encoding.Unmarshal(idBytes, &id)
-	if err != nil {
+	return &pb, nil
```

Comment:
processed block is always handled as a pointer. It's a pretty large object, and you want to make changes to it that persist across function calls.